### ISSUE TITLE
xSQLServerAvailabilityGroupListener fails when using PsDSCRunAsCredential

### ISSUE THREAD

**Issue by gladier** [PowerShell/xSQLServer#460]
**Details of the scenario you try and problem that is occurring:**
Using the example resource configuration below, the resource fails to enumerate anything below SQLSERVER:\\SQL\NODENAME\TEST.

Best I can tell this appears to be UAC related, as the example credentials are a local administrator on the machine and a sysadmin on the SQL instance. I can replicate this issue in a non-elevated PS session logged in as the test user. Elevating the PS session resolves the issue.

Removing the PSDSCRunAsCredential parameter appears to be a fix for this (as this causes the resource to run under the local system context) and I expect the issue actually is from upstream in the SQLPS module. However, the error returned is misleading.
> Trying to make a change to a listener that does not exist.

> $secpasswd = ConvertTo-SecureString “PlainTextPassword” -AsPlainText -Force
> $mycreds = New-Object System.Management.Automation.PSCredential (“username”, $secpasswd)
> xSQLServerAvailabilityGroupListener lstnr {
> InstanceName = "TEST
> AvailabilityGroup = "TESTAG1"
> NodeName = $node.nodename
> Ensure = "Present"
> Name = "TESTAG1L1"
> IpAddress - "192.168.10.45/255.255.252.0"
> Port = "9001"
> DHCP = $false
> **PSDSCRunAsCredential = $mycred**
> }

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
OS: 2012 R2
SQL: 2014 SP3 x64
PS: 5.0

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
6.0.0.0

**Comment by asdglkjkwekdf** [comment]
I have been having a similar issue with this resource.

In my scenario:
* The set works
* The test fails
* Running the resource in a configuration by itself resolves the issue (when debugging the apply stage I copy just the resource I am debugging into a "test" configuration).
* Most issues occurred when xSQLServerAlwaysOnAvailabilityGroup runs just before creating the listener - specifically a best case reproduction required both xSQLServerAlwaysOnAvailabilityGroup  and xSQLServerAvailabilityGroupListener
* I am using a PsDscRunAsCredential which is a SQL setup and install account having sysadmin and high level OS priviledges
* I have tested the path in powershell using my admin account (using the SqlServer module) and have no problems with the path.
* I have run the apply stage with a debugger (using my "test" configuration) and have no problems with the path.
* I am running a custom version with the only change being to Import-SQLPSModule to fix the SqlServer module load issue (I thought there was a report for this but I cannot find it).

To get a working configuration I currently do
xSQLServerAlwaysOnAvailabilityGroup
xSQLServerEndpointPermission
xSQLServerPermission
xWaitforAvailabilityGroup (mostly does not impact reliability)
xSQLServerAvailabilityGroupListener

Basically if I put enough things between xSQLServerAlwaysOnAvailabilityGroup  and xSQLServerAvailabilityGroupListener it seems to work.

I did not need to remove the PsDscRunAsCredential.


**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
 OS: 2016 
 SQL: 2016 SP1 x64
 PS: 5.1

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
 6.0.0.0 - customised fix to Import-SQLPSModule to support both SqlServer and Sqlps

**Comment by Zuldan** [comment]
I'm having the same issue sort of issues. I'm having a very hard time creating a listener. UAC is 100% disabled on the node.

Here is when PSDSCRunAsCredential is not used.

```
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Resource ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-33-48: Testing state of listener TESTAGLISTENER1
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-33-48: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as NT AUTHORITY\SYSTEM
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-33-48: Importing SQLPS module.
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]  in 0.2500 seconds.
Invoke-CimMethod : PowerShell DSC resource MSFT_xSQLServerAvailabilityGroupListener  failed to execute Test-TargetResource functionality with error message: Trying to make a change to a listener that does not exist. 
InnerException: Cannot find path 'SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT\AvailabilityGroups\MyTestAG01\AvailabilityGroupListeners' because it does not exist. 
At C:\Program Files\WindowsPowerShell\Modules\BVISMServerEnforcer\Functions\Public\Invoke-BVEnforcerCheck.ps1:20 char:5
+     Invoke-CimMethod -CimSession $ServerName -Name PerformRequiredCon ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Invoke-CimMethod], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure,Microsoft.Management.Infrastructure.CimCmdlets.InvokeCimMethodCommand
    + PSComputerName        : s12aumdc01tst81
```

Same result when PSDSCRunAsCredential is used.

```
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Resource ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-37-16: Testing state of listener TESTAGLISTENER1
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-37-16: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as LAB\myserviceaccount
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_09-37-16: Importing SQLPS module.
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]  in 0.2970 seconds.
Invoke-CimMethod : PowerShell DSC resource MSFT_xSQLServerAvailabilityGroupListener  failed to execute Test-TargetResource functionality with error message: Trying to make a change to a listener that does not exist. 
InnerException: Cannot find path 'SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT\AvailabilityGroups\MyTestAG01\AvailabilityGroupListeners' because it does not exist. 
At C:\Program Files\WindowsPowerShell\Modules\BVISMServerEnforcer\Functions\Public\Invoke-BVEnforcerCheck.ps1:20 char:5
+     Invoke-CimMethod -CimSession $ServerName -Name PerformRequiredCon ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Invoke-CimMethod], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure,Microsoft.Management.Infrastructure.CimCmdlets.InvokeCimMethodCommand
    + PSComputerName        : s12aumdc01tst81
```

I logged onto the node and had the resource create the listener.

```
$SetParams = @{
    Ensure               = 'Present'
    NodeName             = 'S12AUMDC01TST81'
    InstanceName         = 'MSSQLSERVER'
    AvailabilityGroup    = 'MyTestAG01'
    Name                 = 'TESTAGLISTENER1'
    IpAddress            = "10.137.187.125/255.255.254.0"
    Port                 = 1433
}

MSFT_xSQLServerAvailabilityGroupListener\Set-TargetResource @TestParams -Verbose

VERBOSE: 2017-04-01_09-25-11: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as LAB\myserviceaccount
VERBOSE: 2017-04-01_09-25-11: Importing SQLPS module.
VERBOSE: 2017-04-01_09-25-12: Listener TESTAGLISTENER1 does not exist
VERBOSE: Performing the operation "Create listener on MyTestAG01" on target "TESTAGLISTENER1".
VERBOSE: 2017-04-01_09-25-12: Listener port set to 1433
VERBOSE: 2017-04-01_09-25-12: Listener set to static IP-address(es); 10.80.51.22/255.255.254.0
VERBOSE: 2017-04-01_09-25-28: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as PAC\z-gshipp
VERBOSE: 2017-04-01_09-25-28: Importing SQLPS module.
```

I then ran test to confirm it's there.

```
$TestParams = @{
    Ensure               = 'Present'
    NodeName             = 'S12AUMDC01TST81'
    InstanceName         = 'MSSQLSERVER'
    AvailabilityGroup    = 'MyTestAG01'
    Name                 = 'TESTAGLISTENER1'
    IpAddress            = "10.80.51.22/255.255.254.0"
    Port                 = 1433
}

MSFT_xSQLServerAvailabilityGroupListener\Test-TargetResource @TestParams -Verbose

VERBOSE: 2017-04-01_10-01-42: Testing state of listener TESTAGLISTENER1
VERBOSE: 2017-04-01_10-01-42: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as LAB\myserviceaccount
VERBOSE: 2017-04-01_10-01-42: Importing SQLPS module.
VERBOSE: 2017-04-01_10-01-43: Listener TESTAGLISTENER1 already exist
True
```

I then perform a consistency check and it still fails with the exact errors.

Things I'd like to point out.

- On a previous lab build. I was able to get a consistency check with no errors when I removed PSDSCRunAsCredential. However later on, it failed again so I added PSDSCRunAsCredential back. The consistency check then passed again.

- Using the exact same code I was able to build a second listener called TESTAGLISTENER2 (assigned to a second AG group of course). It was created perfectly and consistency checks always pass for it. It's ONLY the **first** listener that always has an issue. If I rebuild the lab from scratch, the first listener fails again but the second listener is perfect.

Something weird is really going on!

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
OS: 2012
SQL: 2016 SP1 x64
PS: 5.1

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
6.0.0.0

**Comment by Zuldan** [comment]
To add to this. I just rebuilt the lab again. The weird thing is the the first listener was created perfectly.

```
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Resource ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-44: Testing state of listener TESTAGLISTENER1
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-44: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as LAB\myserviceaccount
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-44: Importing SQLPS module.
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Listener TESTAGLISTENER1 does not exist
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]  in 0.7180 seconds.
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Set      ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as PAC\s-enforcer
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Importing SQLPS module.
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Listener TESTAGLISTENER1 does not exist
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] Performing the operation "Create listener on MyTestAG01" on target "TESTAGLISTENER1".
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Listener port set to 1433
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-27-45: Listener set to static IP-address(es); 10.80.51.22/255.255.254.0
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Set      ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]  in 14.8040 seconds.
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Resource ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
```

I then ran a consistency check a few minutes later and the first listener failed but second listener is perfect. Welcome to the world of freakyness!

```
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Resource ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]: LCM:  [ Start  Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-35-54: Testing state of listener TESTAGLISTENER1
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-35-54: Connecting to SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT as LAB\myserviceaccount
VERBOSE: [S12AUMDC01TST81]:                            [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener] 2017-04-01_10-35-54: Importing SQLPS module.
VERBOSE: [S12AUMDC01TST81]: LCM:  [ End    Test     ]  [[xSQLServerAvailabilityGroupListener]SQLAOGroupMyTestAG01Listener]  in 0.5000 seconds.
Invoke-CimMethod : PowerShell DSC resource MSFT_xSQLServerAvailabilityGroupListener  failed to execute Test-TargetResource functionality with error message: Trying to make a change to a listener that does not exist. 
InnerException: Cannot find path 'SQLSERVER:\SQL\S12AUMDC01TST81\DEFAULT\AvailabilityGroups\MyTestAG01\AvailabilityGroupListeners' because it does not exist. 
At C:\Program Files\WindowsPowerShell\Modules\BVISMServerEnforcer\Functions\Public\Invoke-BVEnforcerCheck.ps1:20 char:5
+     Invoke-CimMethod -CimSession $ServerName -Name PerformRequiredCon ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Invoke-CimMethod], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure,Microsoft.Management.Infrastructure.CimCmdlets.InvokeCimMethodCommand
    + PSComputerName        : S12AUMDC01TST81
```

I've tried so many other tests such as using completely different service accounts, IP's and hostname but all have the same issue.

**Comment by johlju** [comment]
I think it is the SQLPS provider that is not stable when using it through DSC? I have had similar (unstable) issues as well, not with this particular resource, but with the SQLPS provider in general when used thru DSC. 

I think one solution is just to remove the dependency on SQLPS provider for this resource, and use the SMO model directly instead, as we do with most other resources.

**Comment by Zuldan** [comment]
> I think it is the SQLPS provider that is not stable when using it through DSC?

@johlju agreed, I don't think the actual resource is the issue, it's something in SQLPS when used with DSC that is causing the instability.

> I think one solution is just to remove the dependency on SQLPS provider for this resource, and use the SMO model directly instead

Sounds like a good idea. Would you have time in the near future to do this conversion? 

> as we do with most other resources.

Standardization is the name of the game ;-)

**Comment by johlju** [comment]
Done and done. Fix is in PR #469. Please report any problems.

I run it in my lab and added and removed a listener without problem.

**Comment by Zuldan** [comment]
@johlju wow that was really fast. Using SMO directly has made it super stable. I can't get it to fail. Love your work.

**Comment by johlju** [comment]
Thank you! Happy it is stable now! :smile:
